### ISSUE TITLE
Degrading Write Speed Performance in Docker with Config "--storage-mode=ON_DISK_TRANSACTIONAL"

### ISSUE THREAD

**Issue by cloudcell** [memgraph/memgraph#2262]
**Memgraph version**
docker image of memgraph:  1.18.1-memgraph-2.18.1

**Environment**
operating system: xubuntu 22.04 LTS jammy jellyfish
architecture: amd x86 64 bit
how do you connect: connecting via 127.0.0.1:57687 (mapped to 7687 in docker to memgraph itself)
with or without docker: with docker (using latest docker desktop v.4.33.0
which driver: using driver (python library "mgclient")

**Describe the bug**
I am importing data into the graph in chunks that have similar sizes. You can see statistics on the import speed: it decreases from 5.67 files per second to less than 1 iteration per second (1.09 seconds per iteration). Please see section "Logs" below.

**To Reproduce**
1. Use the following code to reproduce:
``` 
    def check_relation_exists(self, node1, node2, relation, properties: dict = {}):
        c = ' AND ' + ' AND '.join([f"r.{k} = {v}" if isinstance(v, (int, float)) else f"r.{k} = '{v}'" for k, v in properties.items()]) if properties else ''
        query = f"MATCH (n:Concept)-[r:{relation} * ..1]-(m:Concept) \
                WHERE n.name = '{node1}' AND m.name = '{node2}' {c} \
                RETURN 1 LIMIT 1"
        return bool(self.execute_and_fetch_list(query))

    def add_relation(self, node1, node2, relation, properties: dict = {}):
        properties = '{' + ', '.join([f"{k} : {v}" if isinstance(v, (int, float)) else f"{k} : '{v}'" for k, v in properties.items()]) + '}' if properties else ''
        query = f"MATCH (a:Concept {{name: '{node1}'}}), (b:Concept {{name: '{node2}'}}) \
                MERGE (a)-[r:{relation} {properties}]->(b)"
        self.execute(query)
```
2. create an arbitrary file with arbitrary nodes and relations
3. check if relation exists before submitting a new one using the code provided in step 1
4. see the performance degrade as the size of the graph grows 

**Expected behavior**
The expected speed should be constant (in terms of files/queries processed per second). This is how IN_MEMORY mode works.

**Logs**
Please see below:
```
██| 2142/2142 [06:17<00:00,  5.67it/s]
██| 2058/2058 [12:52<00:00,  2.67it/s]
██| 2089/2089 [18:03<00:00,  1.93it/s]
██| 2073/2073 [23:17<00:00,  1.48it/s]
██| 2096/2096 [28:42<00:00,  1.22it/s]
██| 2134/2134 [34:19<00:00,  1.04it/s]
██| 2064/2064 [37:26<00:00,  1.09s/it]
```

**Additional context**
Some stats on the number of nodes/relations that corresponds to the latest log item (when the performance was already degraded):
![image](https://github.com/user-attachments/assets/b014cda6-c79e-47b1-9f2b-784cb782ea76)

**Verification Environment**
I need the following to verify the fix:
* Plain memgraph package -> for Xubuntu 22.04 LTS jammy jellyfish
* Please provide plain memgraph Docker image.
* Architecture: amd x86?

**Comment by katarinasupe** [comment]
Hi @cloudcell, thank you for reporting this issue! 
Can you tell me the reason why you decided to use on-disk transactional storage mode? I see you mentioned that in-memory mode is performing well for you. Is your import a one-time thing in bulk, or do you plan to update your data on a daily/weekly/monthly basis?  Is this a blocking issue for your project with Memgraph? 
Does the issue you describe happen every time or it depends on some other factors/activities you're doing with Memgraph?

**Comment by cloudcell** [comment]
Hi @katarinasupe, here are the answers to your questions:
* The reason I decided to use on-disk transactional storage mode is that my graph is growing every day.
* The in-memory mode is performing well; however I often need to run several graphs at the same time and docker containers that run them start to fail occasionally due to memory limitations. 
* My import is an ongoing process and I plan to update my data on a monthly and soon daily basis. 
* This is a serious issue because I am considering other alternatives to Memgraph.
* The issue I describe happens every time. The data import performance degrades as the size of the graph grows.

**Comment by katarinasupe** [comment]
How large is the graph @cloudcell? Is there a way to still fit this in memory and use in-memory storage mode?

I understand the experience is not ideal, but as [stated in our docs](https://memgraph.com/docs/fundamentals/storage-memory-usage#on-disk-transactional-storage-mode), the on-disk storage is currently in its experimental phase. Still, we are thankful you opened this issue on GitHub and shared the above information that will be helpful. Improving the on-disk storage mode could become a priority at one point and you can track the work on that on our [roadmap](https://github.com/memgraph/roadmap).

**Comment by cloudcell** [comment]
I understand @katarinasupe , I hope that this will be a viable alternative to other databases in the future.
The graph has over 10 million nodes and over a 120 million relations at this stage (it is expected to double in the next few weeks) and I need to run several instances of this database on the server. Also it would be great if you could enable sharding so the database could be split across multiple servers.

**Comment by katarinasupe** [comment]
I agree that sharding would be a really nice feature. And why do you need several instances of that database on your server? We have successful use cases of graphs bigger than 10M nodes and 120M edges, all running in memory. I get that by having X instances of the same data is X times more RAM, so I was wondering why you even need those extra instances.

**Comment by cloudcell** [comment]
In a couple of days I will have about 20 million nodes with over 200 million relations (estimated roughly). Overall I expect 100 million nodes with perhaps a billion relations in the future (a conservative estimate). I need to run two instances of the database because one is a production instance and the other is a development instance.

**Comment by robcaulk** [comment]
@cloudcell do you mind sharing your expected hardware configuration? We are looking at building a similarly scaled memgraph db.

**Comment by cloudcell** [comment]
Our current hardware is a Supermicro server with half a terabyte of memory. As for the expected one, we are currently working on this. Feel free to DM me on discord on the memgraph server, I would be happy to have a voice chat and discuss this in more detail.

**Comment by katarinasupe** [comment]
If any one of you have more questions, don't hesitate to schedule an [office hours call](https://memgraph.com/office-hours). I appreciate the discussion and thank you @cloudcell for helping out @robcaulk 😄
